// @sentry/node/esm/client.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const os = require("os");
const util = require("util");
const serverRuntimeClient = require('./../_node_modules/@sentry/core/esm/server-runtime-client.cjs');
const version = require('./../_node_modules/@sentry/core/esm/version.cjs');
function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: "Module" } });
  if (e) {
    for (const k in e) {
      if (k !== "default") {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}
const os__namespace = /* @__PURE__ */ _interopNamespaceDefault(os);
class NodeClient extends serverRuntimeClient.ServerRuntimeClient {
  /**
   * Creates a new Node SDK instance.
   * @param options Configuration options for this SDK.
   */
  constructor(options) {
    options._metadata = options._metadata || {};
    options._metadata.sdk = options._metadata.sdk || {
      name: "sentry.javascript.node",
      packages: [
        {
          name: "npm:@sentry/node",
          version: version.SDK_VERSION
        }
      ],
      version: version.SDK_VERSION
    };
    options.transportOptions = {
      textEncoder: new util.TextEncoder(),
      ...options.transportOptions
    };
    const clientOptions = {
      ...options,
      platform: "node",
      runtime: { name: "node", version: global.process.version },
      serverName: options.serverName || global.process.env.SENTRY_NAME || os__namespace.hostname()
    };
    super(clientOptions);
  }
}
exports.NodeClient = NodeClient;
